shared-module/is31fl3741/is31fl3741.c

```diff
@@ -0,0 +1,328 @@
+        self->i2c = NULL;
+    }
+
+    self->base.type = NULL;
+
+    // If a framebuffer was passed in to the constructor, NULL the reference
+    // here so that it will become GC'able
+    self->framebuffer = NULL;
+}
+
+void common_hal_is31fl3741_is31fl3741_set_paused(is31fl3741_is31fl3741_obj_t *self, bool paused) {
+    self->paused = paused;
+}
+
+bool common_hal_is31fl3741_is31fl3741_get_paused(is31fl3741_is31fl3741_obj_t *self) {
+    return self->paused;
+}
+
+void common_hal_is31fl3741_is31fl3741_set_global_current(is31fl3741_is31fl3741_obj_t *self, uint8_t current) {
+    common_hal_displayio_is31fl3741_begin_transaction(self);
+    is31fl3741_set_current(self->i2c, self->device_address, current);
+    common_hal_displayio_is31fl3741_end_transaction(self);
+}
+
+uint8_t common_hal_is31fl3741_is31fl3741_get_global_current(is31fl3741_is31fl3741_obj_t *self) {
+    common_hal_displayio_is31fl3741_begin_transaction(self);
+    uint8_t current = is31fl3741_get_current(self->i2c, self->device_address);
+    common_hal_displayio_is31fl3741_end_transaction(self);
+    return current;
+}
+
+void common_hal_is31fl3741_is31fl3741_refresh(is31fl3741_is31fl3741_obj_t *self, uint8_t *dirtyrows) {
+    common_hal_displayio_is31fl3741_begin_transaction(self);
+
+    uint8_t dirty_row_flags = 0xFF; // only supports 8 rows gotta fix
+    if (dirtyrows != 0) {
+        dirty_row_flags = *dirtyrows;
+    }
+
+    if (!self->paused) {
+        if (self->scale) {
+            // Based on the Arduino IS31FL3741 driver code
+            uint32_t *buffer = self->bufinfo.buf;
+
+            for (int x = 0; x < self->scale_width; x++) {
+                uint32_t *ptr = &buffer[x * 3]; // Entry along top scan line w/x offset
+                for (int y = 0; y < self->scale_height; y++) {
+                    uint16_t rsum = 0, gsum = 0, bsum = 0;
+                    // Inner x/y loops are row-major on purpose (less pointer math)
+                    for (uint8_t yy = 0; yy < 3; yy++) {
+                        for (uint8_t xx = 0; xx < 3; xx++) {
+                            uint32_t rgb = ptr[xx];
+                            rsum += rgb >> 16 & 0xFF;
+                            gsum += (rgb >> 8) & 0xFF;
+                            bsum += rgb & 0xFF;
+                        }
+                        ptr += self->width; // Advance one scan line
+                    }
+                    rsum = rsum / 9;
+                    gsum = gsum / 9;
+                    bsum = bsum / 9;
+                    uint32_t color = 0;
+                    if (self->auto_gamma) {
+                        color = (IS31GammaTable[rsum] << 16) +
+                            (IS31GammaTable[gsum] << 8) +
+                            IS31GammaTable[bsum];
+                    } else {
+                        color = (rsum << 16) + (gsum << 8) + bsum;
+                    }
+                    is31fl3741_draw_pixel(self->i2c, self->device_address, x, y, color);
```

Comment:
How is the speed of this? Would you consider buffering the whole memory and then blasting it all out at once?

Reply:
The speed seems to match the Arduino code. I could play with blasting it all out at once but haven't had a chance to.